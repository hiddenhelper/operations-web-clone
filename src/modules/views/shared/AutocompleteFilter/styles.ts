import { createStyles, makeStyles } from '@material-ui/core/styles';

export const useStyles = makeStyles(theme =>
  createStyles({
    container: {
      backgroundColor: '#fff',

      '& .MuiAutocomplete-inputRoot': {
        '& .MuiAutocomplete-input': {
          padding: '2.5px 4px',
        },
      },
      '& .MuiAutocomplete-endAdornment': {
        display: 'none',
      },
      '& .Mui-focused': {
        border: 'none',
        '& .MuiOutlinedInput-notchedOutline': {
          borderColor: 'transparent',
        },
      },
      '& .MuiInputBase-root': {
        paddingRight: '9px !important',
        height: 39,
        borderBottom: '1px solid #E5E5E5',
        boxShadow: '0 0 5px 0 rgba(0,0,0,0.15)',
        borderRadius: '5px 5px 0 0',
      },
    },
    buttonContainer: {
      '& .MuiAutocomplete-root': {
        '& .MuiFormControl-root': {
          zIndex: 9,
        },
      },
    },
  })
);
