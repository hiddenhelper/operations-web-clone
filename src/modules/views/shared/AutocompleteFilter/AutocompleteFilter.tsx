import React, { memo, useCallback, useState, useEffect } from 'react';

import TextField from '@material-ui/core/TextField';
import Autocomplete from '@material-ui/lab/Autocomplete';
import ClickAwayListener from '@material-ui/core/ClickAwayListener';

import ButtonFilter from '../ButtonFilter';

import { GeneralModel } from '../../../models';
import { useDebounce } from '../../../../utils/useDebounce';
import { noop } from '../../../../utils/generalUtils';
import { useStyles } from './styles';
import { generalGlobalStyles } from '../../../../assets/styles';

export interface IAutocompleteFilterProps {
  value: string;
  label: string;
  optionList: GeneralModel.INamedEntity[];
  isLoading?: boolean;
  autocompleteWidth?: number;
  onChange: (id: string) => void;
  onSearch?: (text: string) => void;
}

const AutocompleteFilter = ({ value, label, optionList, isLoading, autocompleteWidth = 300, onChange, onSearch = noop }: IAutocompleteFilterProps) => {
  const classes = useStyles();
  const generalStyles = generalGlobalStyles();
  const [optionValue, setValue] = useState<GeneralModel.INamedEntity>(null);
  const [search, setSearch] = useState<string>('');
  const debouncedSearch = useDebounce(search, 350);

  const onChangeFilter = useCallback(
    (event, option) => {
      onChange(option?.id);
    },
    [onChange]
  );

  const onChangeSearch = useCallback(
    event => {
      setSearch(event.target.value);
    },
    [setSearch]
  );

  useEffect(() => {
    if (debouncedSearch.length >= 3) onSearch(debouncedSearch);
  }, [debouncedSearch, onSearch]);

  useEffect(() => {
    setValue(optionList.filter(option => option.id !== '').find(option => option.id === value) || null);
  }, [optionList, value]);
  return (
    <ButtonFilter
      text={label}
      transparentButton={true}
      containerStyleClass={classes.buttonContainer}
      render={({ isOpen, handleClose }) =>
        isOpen && (
          <ClickAwayListener onClickAway={handleClose}>
            <Autocomplete
              id="autocomplete-filter"
              data-testid="autocomplete-filter-wrapper"
              style={{ width: autocompleteWidth }}
              value={optionValue}
              disableClearable={true}
              options={optionList}
              getOptionLabel={option => option.name}
              renderOption={option => <span className={generalStyles.textEllipsis}>{option.name}</span>}
              loading={isLoading}
              renderInput={params => (
                <div>
                  <TextField
                    className={classes.container}
                    autoFocus={true}
                    {...params}
                    variant="outlined"
                    type="text"
                    autoComplete="off"
                    placeholder="Search"
                    onChange={onChangeSearch}
                  />
                </div>
              )}
              onChange={onChangeFilter}
              openOnFocus={true}
              open={true}
              disablePortal={true}
            />
          </ClickAwayListener>
        )
      }
    />
  );
};

export default memo(AutocompleteFilter);
