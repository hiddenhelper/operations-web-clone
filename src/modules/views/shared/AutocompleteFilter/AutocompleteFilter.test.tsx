import React from 'react';
import { render, act, fireEvent } from '@testing-library/react';

import AutocompleteFilter, { IAutocompleteFilterProps } from './AutocompleteFilter';

jest.useFakeTimers();

describe('AutocompleteFilter Component', () => {
  let props: IAutocompleteFilterProps;

  beforeEach(() => {
    props = {
      value: '',
      label: 'Label',
      optionList: [{ id: 'id', name: 'name' }],
      onChange: jest.fn(),
      onSearch: jest.fn(),
    };
  });

  it('should render', () => {
    const wrapper = render(<AutocompleteFilter {...props} />);
    expect(wrapper.container).toMatchSnapshot();
  });

  it('should change', () => {
    const wrapper = render(<AutocompleteFilter {...props} />);

    act(() => {
      fireEvent.click(wrapper.getByTestId('button-filter-open'));
    });

    const input = wrapper.getByTestId('autocomplete-filter-wrapper').querySelector('input');

    act(() => {
      fireEvent.change(input, { target: { value: 'name' } });
    });

    act(() => {
      jest.runOnlyPendingTimers();
    });

    act(() => {
      fireEvent.click(wrapper.getByText('name'));
    });

    expect(props.onChange).toHaveBeenCalledWith('id');
    expect(props.onSearch).toHaveBeenCalledWith('name');
  });
});
