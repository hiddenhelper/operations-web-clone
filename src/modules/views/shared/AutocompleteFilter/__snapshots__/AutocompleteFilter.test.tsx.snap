// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutocompleteFilter Component should render 1`] = `
<div>
  <div
    class="makeStyles-floatingFilterWrapper-26 makeStyles-filterWrapper-25 makeStyles-buttonContainer-2 floatingFilterWrapper"
  >
    <span
      class="makeStyles-floatingFilterButton-27"
      data-testid="controlled-button"
      tabindex="-1"
    >
      <button
        aria-controls="false"
        aria-haspopup="true"
        class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-buttonFilter-35"
        data-testid="button-filter-open"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Label
          <span
            class="MuiButton-endIcon MuiButton-iconSizeMedium"
          >
            <svg
              class="makeStyles-dropdownIcon-29"
            >
              expand.svg
            </svg>
          </span>
        </span>
      </button>
    </span>
  </div>
</div>
`;
